src/Sparrow/Collections/RingBuffer.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Sparrow.Collections
+{
+    // The RingItem makes sure that it doesn't matter how big the struct for the T, it is always guaranteed that the data
+    // will be always on an L0 cache line different from the item that comes before. 
+    [StructLayout(LayoutKind.Sequential)]
+    public struct RingItem<T>
+    {
+        private readonly long _p1, _p2, _p3, _p4;
+
+        internal bool IsReady;
+        public T Item;
+
+        private readonly long _p5, _p6, _p7, _p8;
+    }
+
+    public sealed class SingleConsumerRingBuffer<T>
+    {
+        private readonly RingItem<T>[] _buffer;
+
+        private readonly int _size;
+
+        private int _startIdx;
+        private int _acquiredIdx;
+        private int _currentIdx;
+
+        public SingleConsumerRingBuffer(int size)
+        {
+            // PERF: If we 'limit' size to be power of 2, we can make all operations much more efficient without using idiv operations.
+            this._size = size;
+            this._buffer = new RingItem<T>[size];
+
+            this._startIdx = 0;
+            this._acquiredIdx = -1;
+            this._currentIdx = 0;
+        }
+
+        public bool TryPush(ref T item)
+        {
+            // We will check if we potentially have the ability to push 1 more item. 
+            int cidx = Volatile.Read(ref _currentIdx);
+            int sidx = Volatile.Read(ref _startIdx);
+
+            // Check if we have empty spaces in the buffer.
+            if (cidx - sidx >= this._size)
+                return false; // No space to do anything
+
+            int ticker = Interlocked.Increment(ref _currentIdx) - 1;
+
+            // Assign the value
+            ref var cl = ref _buffer[ticker % this._size];
```

Comment:
The compiler can't assume power of 2 here, no?
Can we turn this to a shift?

Reply:
We can if we control the size. Easy change if we want to do so.